Make Sprockets more optional, offer Propshaft as alternative

dhh opened the pull request:
See more on [rails/propshaft](https://github.com/rails/propshaft).

- [ ] Add explicit sprockets gem dependency to app Gemfile when upgrading existing apps.

Commit:
Make Sprockets more optional, offer Propshaft as alternative

Commit:
Whups, local reference

Commit:
No longer used

Commit:
Spacing

Commit:
Need explicit sprockets-rails inclusion now

Commit:
Manually require the sprockets railtie

Commit:
Don't need these changes right now

Commit:
Merge branch 'main' into support-propshaft

* main:
  [ci skip] Add Bootstrap and Bulma to the CSS processors' list (#43254)
  Add missing configuration to middleware test
  Add ActionDispatch::ServerTiming to the list of middlewares in test
  Add CHANGELOG entry for ServerTiming middleware [ci skip]
  Add Server Timing middleware (#36289)
  Don't overwrite default opts in rich_text_area_tag (#43156)
  Only call `wrap_parameters` if it's defined (#43259)
  Suggest a CSP that's compatible with Turbo + import map (#43227)
  Fix app generator tests
  Fix docs spacing
  Fix typo: integer numbers (not integral)
  Add `beginning_of_week` option to `weekday_options_for_select`
  Pluralize the heading on the index page generated with scaffold
  Move the parameter wrapper code to the ActionController::Railtie class
  Add explicit rendering to DiskController#update
  Fix update & destroy queries when default_scope is nillable with all_queries: true

Commit:
Kick off another build

Commit:
Fix tests

Commit:
DRY up test

Commit:
Require railtie when using sprockets

Commit:
Introduce option to skip asset pipeline

Commit:
No longer relevant

Commit:
Always have to return

Commit:
Gone

Commit:
Add helper for skip_sprockets?

rafaelfranca commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -163,6 +164,10 @@ def config_when_updating
       end
+
+      if uses_sprockets
+        insert_into_file "config/application.rb", %(require "sprockets/railtie"), after: /require\(["']rails\/all["']\)\n/
```
`rails\/all` is not always present in this file. At this point, if `skip-activestorage` is passed for example it will not be there.

rafaelfranca replied:
Ah, another alternative is changing `sprockets-rails` to require the railtie, that way we don't need it in `application.rb` anymore.

dhh replied:
That would actually be way nicer. Since that’s how normal rails integration gems work. Will there be a backwards compatibility issue though?

rafaelfranca replied:
It would change a little bit when it is loaded, but that should not be a problem.

It would be required at this line https://github.com/rails/rails/blob/20db0845cddf6e90a82a782ef51196b9ecb3716d/railties/lib/rails/generators/rails/app/templates/config/application.rb.tt#L24.

Instead of this line https://github.com/rails/rails/blob/20db0845cddf6e90a82a782ef51196b9ecb3716d/railties/lib/rails/generators/rails/app/templates/config/application.rb.tt#L18

But, if an existing app still has the require, or is requiring `'rails/all'`, the second require would just return `false`, so no behavior would change.

dhh replied:
Ah nice. Let’s make it so.

rafaelfranca replied:
Ah, we already do it https://github.com/rails/sprockets-rails/blob/master/lib/sprockets/rails.rb#L3.

So all we need to do is to remove the requires and move to the gem to the Gemfile

Commit:
Fix guard statement

Commit:
Use latest gems

Commit:
Include propshaft

skipkayhil committed:
fix tests for #43261 (#43277)

* help fix tests for #43261

skip_sprockets? should not be called on options

:skip_sprockets is no longer a value in the option hash, so
skip_sprockets? should not be called on it

move --asset-pipeline to shared generator

skip_sprockets? is defined on app_base, and used in the plugin
generator to determine whether to add the engine's assets to the dummy
sprockets manifest, so I believe it makes sense to include in both
generators

Because of this change, I also changed the shared test back to testing
against non-sprockets

add skip_sprockets to Gemfile template vars

Mocking skip_sprockets? in app_base generator

fix more generator tests

* use skip_sprockets? everywhere

Commit:
Use latest propshaft

rmacklin commented:
Opened a PR into this branch with a minor docs update: https://github.com/rails/rails/pull/43328

rmacklin committed:
Update `AssetUrlHelper` docs to list both asset pipeline gems (#43328)

SkipKayhil commented:
@dhh The cyclic dependency error was caused by propshaft's `append_asset_path` initializer referencing itself:
```
#<Rails::Initializable::Initializer:0x00007fdfd42f6ed8
    @block=#<Proc:0x00007fdfd42c6af8 /Users/hartley/.cache/asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/propshaft-0.1.4/lib/propshaft/railtie.rb:7>,
    @context=#<ActionMailbox::Engine:0x00007fdfd7955bf0 ...>,
    @name=:append_assets_path,
    @options={:group=>:all, :after=>:append_assets_path}>
```
This happened because `:after` defaults to `initializers.last.name`, and the last initializer created before propshaft's was `append_assets_path` in sprockets-rails' railtie.

The duplicate initializer name was resolved in https://github.com/rails/propshaft/commit/e07841a8f1b4e16d60393082996e1177152aca81 so I believe a new release of propshaft should fix the tests. I also ran the tests locally with `gem 'propshaft', github: 'rails/propshaft'` and that made them pass.

Commit:
Update to latest Propshaft

Commit:
Merge branch 'main' into support-propshaft

* main: (78 commits)
  Update MessageEncryptor guide to show that an exception is raised
  Depend on Zeitwerk 2.5.0.beta5
  Fix wrong RDoc
  Add `partial_inserts` default configuration for Rails 7.0 in guides/source/configuring.md
  Improves ActionView FormHelper check_box documentation
  Update .github/pull_request_template.md
  Warning about draft PRs
  Add changelog entry for https://github.com/rails/rails/pull/42501
  Fix codespell `hiearchy ==> hierarchy` error.
  Replace the word dumb with naive for clarity and inclusivity
  More example in the docs about STI and autoloading
  Client ensures subscribe command is confirmed. (#41581)
  Document easier way to preload STIs
  Revert "Fix ForkTracker on ruby <= 2.5.3"
  Don't override Object#fork in Ruby 3.0
  Fix default scope with all_queries on reload test name
  Use regular memoization
  Avoid class_attrs for unused collection callbacks
  Accept optional transaction args to #with_lock
  Append trailing slash when the path is generated
  ...

Commit:
Bump Propshaft again

Commit:
Ask for latest

Commit:
Merge branch 'main' into support-propshaft

* main: (21 commits)
  Don't require role when passing shard to connected_to
  Replace ableist language
  Move AllMessages behaviour into Matcher
  Remove "matches?" from AS::N subscriber classes
  Add ability to lazily load the schema cache on connection
  Remove message from ActiveRecord::Rollback example
  Use queue_classic branch which works on psql 14
  Restore set_autoload_path triggering before bootstrap
  Specify ORDER BY enumsortorder for postgres enums
  Automatically infer inverse_of with scopes
  Remove unnecessary if in ExtendedDeterministicUniquenessValidator
  Replace "ActionText" with "Action Text" [ci skip]
  Fix new method name in DatabaseConfig#config deprecation message
  Remove unused instrumentation hooks from action_controller
  Update the documentation for find_by method
  Change "bash" to "shell" in getting started guide
  Clarify try calls public methods only
  Fix invalid link to List of Free Programming Books
  Update docs for nested_attributes to reflect default belongs_to behavior and validating parent model.
  Fix typo
  ...

Commit:
Merge branch 'main' into support-propshaft

* main:
  Revert STI samples in autoloading guide
  Remove refenrence to destroy_all_in_batches config
  Replace more ableist language
  Ensure we guard against all exceptions, not just StandardError
  Update tests to support guarded InstrumentationSubscriberError
  Guard against subscriber exceptions in ActiveSupport::Notifications

Commit:
Use latest propshaft
